### PR TITLE
Update ComponentGraph `_validate_component_dict` logic

### PR DESCRIPTION
**Pull request by angela97lin**
- Remove "Cannot have multiple y parents for a single component" check in _compute_features, all checking is done in `_validate_component_dict` now.
- Cleaned up and added edge case to `_validate_component_dict` (where user specifies .x, .y, but also "thisisnotvalid" 😂). Before, this check would pass as long as there was one .x, one .y properly. This update ensures that all specified values must be .x/.y/X, y :)

### CODE REVIEW

File: evalml/pipelines/component_graph.py [hunk 1]
```diff
@@ -61,13 +61,35 @@ def _validate_component_dict(self):
                 component_input.endswith(".x") or component_input == "X"
                 for component_input in component_inputs
             )
-            has_target_input = any(
+            has_one_target_input = sum(
                 component_input.endswith(".y") or component_input == "y"
                 for component_input in component_inputs
             )
-            if not (has_feature_input and has_target_input):
+            if not has_feature_input:
```
Comment (angela97lin) [inline evalml/pipelines/component_graph.py:68]: Kind of redundant / can be covered by case below. Kept for the specific ValueError, but open to just removing and having it encapsulated below 😂

File: evalml/tests/pipeline_tests/test_component_graph.py [hunk 2]
```diff
@@ -1917,9 +1914,35 @@ def test_component_graph_does_not_define_all_edges():
...
 
 
+def test_component_graph_defines_edges_with_bad_syntax():
+    # Graph does not define an X edge
+    with pytest.raises(
+        ValueError, match="All edges must be specified as either an input feature"
+    ):
+        ComponentGraph(
+            {
+                "Imputer": [Imputer, "X", "y"],  # offending line
```
Comment (freddyaboulton) [inline evalml/tests/pipeline_tests/test_component_graph.py:1924]: This is not actually the offending line right?

File: evalml/tests/pipeline_tests/test_component_graph.py [hunk 3]
```diff
@@ -1917,9 +1914,35 @@ def test_component_graph_does_not_define_all_edges():
...
+                "One Hot Encoder": [OneHotEncoder, "Imputer.x", "y"],
+                "Target Imputer": [
+                    TargetImputer,
+                    "Imputer",
+                    "One Hot Encoder.x",
+                    "y",
+                ],  # offending line: Imputer not allowed
```
Comment (freddyaboulton) [inline evalml/tests/pipeline_tests/test_component_graph.py:1931]: nit-pick: I think the comment would make more sense if it was above "Target Imputer"

File: evalml/tests/pipeline_tests/test_component_graph.py [hunk 4]
```diff
@@ -700,7 +693,7 @@ def test_component_graph_evaluation_plumbing(
     graph = {
         "transformer a": [TransformerA, "X", "y"],
         "transformer b": [TransformerB, "transformer a.x", "y"],
-        "transformer c": [TransformerC, "transformer a", "transformer b.x", "y"],
+        "transformer c": [TransformerC, "transformer a.x", "transformer b.x", "y"],
```
Comment (freddyaboulton) [inline evalml/tests/pipeline_tests/test_component_graph.py:696]: Ok so before "transformer a" was being silently ignored? I'm curious why this change doesn't mean we also have to change the actual expected call args for `mock_transc`

**Review by freddyaboulton — APPROVED** [review]
Thank you @angela97lin !!

File: evalml/tests/pipeline_tests/test_component_graph.py (hunk 4 above)
Comment (freddyaboulton) [reply evalml/tests/pipeline_tests/test_component_graph.py:696]: Ohh it's cause the output of transformer b already had the output of transformer a. I wonder if we should change the test so that the column names don't overlap? Duplicate column names would cause ww to error out anyways.

File: evalml/tests/pipeline_tests/test_component_graph.py (hunk 2 above)
Comment (angela97lin) [reply evalml/tests/pipeline_tests/test_component_graph.py:1924]: Wow I swore I updated this but I must have dreamed it. Thank you 😂

File: evalml/tests/pipeline_tests/test_component_graph.py (hunk 3 above)
Comment (angela97lin) [reply evalml/tests/pipeline_tests/test_component_graph.py:1931]: Going to put it next to the "Imputer" :)

File: evalml/tests/pipeline_tests/test_component_graph.py (hunk 4 above)
Comment (angela97lin) [reply evalml/tests/pipeline_tests/test_component_graph.py:696]: Hehe so right now, `parent_x = output_cache.get(parent_input, output_cache.get(f"{parent_input}.x"))` means that it'll try either getting `transformer a` or `transformer a.x` as a backup, which is why this worked. We're no longer going to have to default to this anymore but but https://github.com/alteryx/evalml/pull/2583 is going to clean that up further :)
